### PR TITLE
Add completion index as pod label for indexed jobs

### PR DESCRIPTION
**Pull request by danielvegamyhre**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Implementing Indexed Job changes for [KEP-4017](https://github.com/kubernetes/enhancements/pull/4019).

#### Which issue(s) this PR fixes:
Part of https://github.com/kubernetes/enhancements/issues/4017

#### Does this PR introduce a user-facing change?
Yes, the completion index of a pod will now be added as a pod label for indexed jobs.

```release-note
Indexed Job pods now have the pod completion index set as a pod label.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
[KEP](https://github.com/kubernetes/enhancements/pull/4019)
```

### CODE REVIEW

**Comment by danielvegamyhre** [comment]
tagging @alculquicondor for review. let me know if you think we need any additional testing.

**Review by kannon92 — COMMENTED** [review]
Rereading your KEP I think you need to add a feature toggle for this work.

File: pkg/controller/job/job_controller_test.go [hunk 1]
```diff
@@ -4394,6 +4395,13 @@ func TestFinalizersRemovedExpectations(t *testing.T) {
 		t.Errorf("Timeout waiting for expectations (-want, +got):\n%s", diff)
 	}
 }
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:4397]: add empty line after this

**Review by alculquicondor — COMMENTED** [review]
Yes, this need a feature gate, even if it's enabled by default.

**Comment by danielvegamyhre** [comment]
> Rereading your KEP I think you need to add a feature toggle for this work.

Thanks for the reminder! Added a feature gate.

File: pkg/controller/job/job_controller_test.go (hunk 1 above)
Comment (danielvegamyhre) [reply pkg/controller/job/job_controller_test.go:4399]: done

**Comment by alculquicondor** [comment]
tentatively
/assign @thockin
as API reviewer, as he has already shared some thoughts about annotations

File: pkg/features/kube_features.go [hunk 2]
```diff
@@ -843,6 +843,13 @@ const (
 	//
 	// Enables In-Place Pod Vertical Scaling
 	InPlacePodVerticalScaling featuregate.Feature = "InPlacePodVerticalScaling"
+
+	// owner: @danielvegamyhre
+	// kep: https://kep.k8s.io/4017
+	// beta: v1.28
+	//
+	// Set pod completion index as a pod label for Indexed Jobs and StatefulSets.
+	PodIndexLabel featuregate.Feature = "PodIndexLabel"
```
Comment (thockin) [inline pkg/features/kube_features.go:852]: These lists are supposed to be alphabetized.  Can you insert it in the right place?  I know not everything is correct.

**Comment by thockin** [comment]
What does "controller" or "batch" add to the clarity here?   It doesn't tell me anything extra.  Do we think this annotation is exclusively used by batch?  There have been discussions about `IndexedDeployment` in the past - would we want that to use the same annotation, or a different one?

`pod.kubernetes.io/index`  or  `kubernetes.io/pod-index`   seem like they encode the right number of facts.  But it asks another question - index into what?  At least with Deployments, a pod can be orphaned and adopted by a different Deployment.  Is index guaranteed unique, or can I have multiple replicas of a given index?  Is it possible to select all the things that have indexes in the same set?  How?

**Comment by thockin** [comment]
@smarterclayton since we discussed

File: pkg/features/kube_features.go (hunk 2 above)
Comment (danielvegamyhre) [reply pkg/features/kube_features.go]: Sure, done.

**Comment by danielvegamyhre** [comment]
> `pod.kubernetes.io/index` or `kubernetes.io/pod-index` seem like they encode the right number of facts. 

`pod.kubernetes.io/index` or `kubernetes.io/pod-index` are both fine with me. You're suggesting this name solely for the new label, correct? Since I think migrating the existing annotation `batch.kubernetes.io/job-completion-index` to the new name as well would be a breaking change.

> Is index guaranteed unique, or can I have multiple replicas of a given index? Is it possible to select all the things that have indexes in the same set? How?

I think the more specific name of `batch.kubernetes.io/job-completion-index` answers these questions, since job completion indexes are a feature of indexed jobs with documented behavior. However, I'm open to either naming convention. @ahg-g any thoughts on this?

**Comment by ahg-g** [comment]
> But it asks another question - index into what? 

The replicas that belong to the same owner.

> Is index guaranteed unique, or can I have multiple replicas of a given index? 

This is a property of the app, it is unique for statefulset and indexed job. We could document that any app that wishes to use this label should ensure that the index is unique (and so perhaps we name it pod-unique-index) or we could leave it open and say uniqueness is inherited from the app. I prefer the latter.

> Is it possible to select all the things that have indexes in the same set? How?

I am not following, can you clarify the question? it is a label, so whatever applies to labels applies to the index.

**Comment by ahg-g** [comment]
> Is it possible to select all the things that have indexes in the same set? How?

I see now what you mean, there is no common label for owner uid or name; that is why I suggested we add statefulset-name and job-name labels as well; if we are going to have a common label for the index, we could have a common label for the owner name (perhaps the api-server can do that for all pods by copying the name in the ownerReference to a label);

**Comment by soltysh** [comment]
> `pod.kubernetes.io/index` or `kubernetes.io/pod-index` seem like they encode the right number of facts. But it asks another question - index into what? At least with Deployments, a pod can be orphaned and adopted by a different Deployment. Is index guaranteed unique, or can I have multiple replicas of a given index? Is it possible to select all the things that have indexes in the same set? How?

@thockin that's why we were discussing being more explicit with the purpose here, currently we have `statefulset.kubernetes.io/pod-name` which exposes information about pod name in a statefulset, in parallel to that we also have `batch.kubernetes.io/job-completion-index`. The purpose is similar for both, but the information is presented in non-unified way. Based on your comment maybe we could make it explicitly as `<controller-name>.kubernetes.io/<name>` to clearly articulate both the ownership through the first part, and the purpose of the label. 

I've raised this topic while reviewing the KEP that we have some inconsistencies, but it would be nice to have a unified way which would clearly tell users what every annotation expresses. wdyt?

**Comment by ahg-g** [comment]
> > Is it possible to select all the things that have indexes in the same set? How?
> 
> Yes, using the labelSelector defined for the Job/Statefulset

That is not too convenient, especially for job where the selector is added automatically and is based on the job's uid. Looking up the pods using the job name is more intuitive.

File: pkg/controller/job/job_controller_test.go [hunk 3]
```diff
@@ -962,6 +962,9 @@ func checkIndexedJobPods(t *testing.T, control *controller.FakePodControl, wantI
 	gotIndexes := sets.New[int]()
 	for _, p := range control.Templates {
 		checkJobCompletionEnvVariable(t, &p.Spec)
+		if feature.DefaultFeatureGate.Enabled(features.PodIndexLabel) {
```
Comment (kannon92) [inline pkg/controller/job/job_controller_test.go:965]: The pattern is usually to explicity enable the unit test as a parameter.

I would pass in this feature toggle as a test case parameter and verify the negative and positive case.  

I know this is default enabled but you still want to test that this works if an admin turned this off.

**Comment by thockin** [comment]
> You're suggesting this name solely for the new label, correct? 

Right - whatever old names you set have to still be set.

> I think the more specific name of batch.kubernetes.io/job-completion-index answers these questions, since job completion indexes are a feature of indexed jobs with documented behavior

> Based on your comment maybe we could make it explicitly as `<controller-name>.kubernetes.io/<name>` to clearly articulate both the ownership

This whole KEP is about defining a common name, is it not?  Controller-name is obviously more specific, but is that the intention?  I like the idea of metadata that is abstract enough to not encode the specific provenance of the pod.  The question I have is whether making it abstract reduces the value of it.  If it is PURELY a way of publishing information down to the pod, we maybe only need to assert some semantics and then make controllers implement those. 

> That's why I prefer controller.kubernetes.io/pod-index

I don't see that "controller" adds anything.  It doesn't say which controller.  It doesn't allow for multiple controllers or to have different indices.  (e.g. you are the 3rd replica of shard 9).

> Actually we add batch.kubernetes.io/job-name label for job, but there is nothing similar for statefulset

I have an idea for a more generic "group ID" sort of label, but I need to think through propagation rules.  The problem with "job-name" or "deployment name" is that (at least for Deployments) it can change at runtime, so anything you decorate with that information has to be adaptable or has to be willing to be wrong.

So let's enumerate semantics:
* Numeric
* Within some larger but ill-defined (for now) set
* Not necessarily unique in set or in time
* Is it supposed to be mutable?  What does it mean if it is changed at runtime?
* Should controllers re-assert this value if someone else were to change it (e.g. a human messing around)?
* Is it always zero-based?  I think there's an ordinal offset field, right?  So no, at least not within a cluster.
* Is it allowed to be sparse (e.g. use 0 and 5, but not 2, 3, 4)?
* Would you use this for data-shard assignments to replicas in a Deployment?

File: pkg/controller/job/job_controller_test.go (hunk 3 above)
Comment (danielvegamyhre) [reply pkg/controller/job/job_controller_test.go:977]: Implemented your suggestion, let me know how it looks.

Since the feature flag is default enabled, i opted to invert the usual parameter pattern `<featureflag>Enabled bool` flag to `<featureflag>Disabled bool`, so that all the existing test cases that don't specify this parameter will execute with the feature flag enabled, and then I added one test case which sets this parameter to true to execute indexed job test case with the feature flag disabled. This way we achieve full unit test coverage for a "default enabled" feature which I think is important. 

What do you think? Does this approach make sense or is it there a better alternative?

**Comment by ahg-g** [comment]
> Numeric
> Not necessarily unique in set or in time

+1

> Within some larger but ill-defined (for now) set

I would say share the same owner.

> Is it supposed to be mutable? What does it mean if it is changed at runtime?
> Should controllers re-assert this value if someone else were to change it (e.g. a human messing around)?

I would say it is set at pod creation only, and not necessarily reconciled.

> Is it always zero-based? I think there's an ordinal offset field, right? So no, at least not within a cluster.
> Is it allowed to be sparse (e.g. use 0 and 5, but not 2, 3, 4)?

This depends on the controller, so to be flexible, we can say it is not necessarily zero-based and it is allowed to be sparse. 

> Would you use this for data-shard assignments to replicas in a Deployment?

This is not supported in Deployment, but we do recommend to use the index to statically assign a partition of the data for each worker in StatefulSet and Indexed Job.

**Comment by thockin** [comment]
>> Is it supposed to be mutable? What does it mean if it is changed at runtime?
>> Should controllers re-assert this value if someone else were to change it (e.g. a human messing around)?
>
> I would say it is set at pod creation only, and not necessarily reconciled.

Labels and annotations are a poor fit for this.  They ARE mutable, so we need to define the semantics.
Is this literally set AT CREATION or is it set some time AT OR AFTER creation?  IOW, could I have a shard-assigner that used this to publish assignments?  What do we expect to happen if a user or other controller changes or deletes this annotation?

What if this were a field on Pod?  Would that make more sense?  Not really proposing this, but interesting to think about.

File: pkg/controller/job/job_controller_test.go (hunk 3 above)
Comment (danielvegamyhre) [reply pkg/controller/job/job_controller_test.go:977]: Just realized I forgot to push the changes described in my previous comment 😅 Just pushed the changes. Let me know what you think.

Comment (kannon92) [reply pkg/controller/job/job_controller_test.go:977]: TBH I don't really understand how feature flags work in k/k but I don't see a lot of test code that uses this pattern.  

I think you should just check the flag you pass in rather than the feature toggle as this seems different from the rest of the code.

```
if !tc.podIndexLabelDisabled {...}
```

File: pkg/features/kube_features.go [hunk 4]
```diff
@@ -606,6 +606,13 @@ const (
 	// the pod is being deleted due to a disruption.
 	PodDisruptionConditions featuregate.Feature = "PodDisruptionConditions"
 
+	// owner: @danielvegamyhre
+	// kep: https://kep.k8s.io/4017
+	// beta: v1.28
+	//
+	// Set pod completion index as a pod label for Indexed Jobs and StatefulSets.
```
Comment (kannon92) [inline pkg/features/kube_features.go:613]:
```suggestion
	// Set pod completion index as a pod label for Indexed Jobs.
```

I imagine you are going to do SS in another PR but the comment is incorrect currently.

**Comment by ahg-g** [comment]
> Is this literally set AT CREATION or is it set some time AT OR AFTER creation? IOW, could I have a shard-assigner that used this to publish assignments? What do we expect to happen if a user or other controller changes or deletes this annotation?
> 

It is literally set at creation, not after.

> Labels and annotations are a poor fit for this. They ARE mutable, so we need to define the semantics. 
> What if this were a field on Pod? Would that make more sense? Not really proposing this, but interesting to think about.

Yeah, we do need them as labels so we can target them using label selectors.

File: pkg/features/kube_features.go (hunk 4 above)
Comment (danielvegamyhre) [reply pkg/features/kube_features.go]: Updated

**Comment by thockin** [comment]
Revisiting this:

Using a label so it is selectable seems reasonable.  So now we discuss the name.

Very specific names are most clear, but given that we have at least two cases that have very similar semantics, the obvious question is whether to use the same name for them both (and thereby define some set of semantics for that name beyond these two).

Two is not a very big number from which to extrapolate commonalities.  So, while I initially thought it would be "neat" to define a generic label, I am not sure I see the value that being generic would bring, while I do see the risk of over-generalizing.

So, unless there *is* a strong motivation to use the same name (in which case, please lay that out), I propose that you define clear, controller/apigroup/whatever specific names, and we can revisit the need for an additional label at some point later.

What do you think?

**Comment by danielvegamyhre** [comment]
> Two is not a very big number from which to extrapolate commonalities. So, while I initially thought it would be "neat" to define a generic label, I am not sure I see the value that being generic would bring, while I do see the risk of over-generalizing.

+1 on this. 

> So, unless there is a strong motivation to use the same name (in which case, please lay that out), I propose that you define clear, controller/apigroup/whatever specific names, and we can revisit the need for an additional label at some point later.

I think `<controller-name>.kubernetes.io/pod-index` makes sense, since using the controller name is more precise than the API group name, and if we are not going down the generic path, more specificity seems like a good thing (i.e. `statefulset.kubernetes.io/pod-index` and `job.kubernetes.io/pod-index` are more clear than rather than `apps.kubernetes.io/pod-index` and `batch.kubernetes.io/pod-index`). 

I'm curious to hear others thoughts on this as well.

**Comment by ahg-g** [comment]
Thanks Tim, I think continuing with the original plan of having separate labels is a good path. 

This exploration on unifying the key was well worth it though and we should capture it in the KEP. @danielvegamyhre can you please do that? it is not urgent to do now since we didn't change the plan or record, but I think we should document this under alternative solutions.

**Comment by danielvegamyhre** [comment]
> Thanks Tim, I think continuing with the original plan of having separate labels is a good path.
> 
> This exploration on unifying the key was well worth it though and we should capture it in the KEP. @danielvegamyhre can you please do that? it is not urgent to do now since we didn't change the plan of record, but I think we should document this under alternative solutions.

Sounds good, will do.

**Comment by danielvegamyhre** [comment]
Ok tests are passing and this is ready for review. Please note this feature flag is enabled by default.

File: pkg/controller/job/job_controller_test.go [hunk 5]
```diff
@@ -274,7 +274,8 @@ func TestControllerSyncJob(t *testing.T) {
 		expectedPodPatches      int
 
 		// features
-		jobReadyPodsEnabled bool
+		jobReadyPodsEnabled   bool
+		podIndexLabelDisabled bool
```
Comment (danielvegamyhre) [inline pkg/controller/job/job_controller_test.go:278]: Reviewer note: Since the feature flag is default enabled, I inverted the usual parameter pattern `<featureflag>Enabled bool` flag to `<featureflag>Disabled bool`, so that all the existing test cases that don't specify this parameter will execute with the feature flag enabled, and then I added one test case which sets this parameter to true to execute indexed job test case with the feature flag disabled. This way we achieve full unit test coverage for a "default enabled" feature (which I think is important), while still having some verification that things work when it is disabled.

**Comment by soltysh** [comment]
+1 to the idea of `<controller-name>.kubernetes.io/pod-index` so here it's `job.kubernetes.io/...` which is what I originally meant, but poorly expressed :sweat_smile:

**Comment by alculquicondor** [comment]
To @deads2k's point, we already have a significant number of {labels, annotations, finalizers} with prefix `batch.kubernetes.io`.
Furthermore, we initiated a deprecation for plain [`job-name`](https://kubernetes.io/docs/reference/labels-annotations-taints/#job-name), in favor of `batch.kubernetes.io/job-name` in 1.27.
It could be annoying for users if deprecate these new labels so soon.

File: staging/src/k8s.io/api/batch/v1/types.go [hunk 6]
```diff
@@ -28,6 +28,8 @@ const (
 	labelPrefix = "batch.kubernetes.io/"
 
 	JobCompletionIndexAnnotation = labelPrefix + "job-completion-index"
+	// JobCompletionIndexLabel is defined following the format <controller>.kubernetes.io/pod-index
```
Comment (kannon92) [inline staging/src/k8s.io/api/batch/v1/types.go:31]: I think you need to replicate this here:

https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/batch/types.go#L25

Comment (danielvegamyhre) [reply staging/src/k8s.io/api/batch/v1/types.go:31]: I don't see the JobCompletionIndexAnnotation there though, seems like it doesn't include everything, do you know why that is?

File: pkg/controller/job/job_controller_test.go (hunk 5 above)
Comment (kannon92) [reply pkg/controller/job/job_controller_test.go:278]: Thanks for adding that!  I think it looks fine to me.  And thank you for the note.  It is a bit odd pattern but I can understand why you did it that one.

File: staging/src/k8s.io/api/batch/v1/types.go (hunk 6 above)
Comment (ahg-g) [reply staging/src/k8s.io/api/batch/v1/types.go:31]: It is not necessary to put it in the internal api file unless the api-server needs to reference it.

File: pkg/controller/job/job_controller.go [hunk 7]
```diff
@@ -1482,6 +1482,10 @@ func (jm *Controller) manageJob(ctx context.Context, job *batch.Job, activePods
 					if completionIndex != unknownCompletionIndex {
 						template = podTemplate.DeepCopy()
 						addCompletionIndexAnnotation(template, completionIndex)
+
+						if feature.DefaultFeatureGate.Enabled(features.PodIndexLabel) {
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:1486]: Also change `addCompletionIndexEnvVariable` to use the label instead of the annotation.

This way, we can deprecate the annotation and remove it in the future.

File: staging/src/k8s.io/api/batch/v1/types.go [hunk 8]
```diff
@@ -28,6 +28,8 @@ const (
...
+	JobCompletionIndexLabel = "job.kubernetes.io/pod-index"
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply staging/src/k8s.io/api/batch/v1/types.go:32]: we are settling on the prefix `batch.kubernetes.io`.

I'm ok with either `/job-completion-index` or `/job-pod-index`.

File: pkg/controller/job/job_controller.go (hunk 7 above)
Comment (danielvegamyhre) [reply pkg/controller/job/job_controller.go:1486]: Done

**Comment by danielvegamyhre** [comment]
@alculquicondor I addressed your comments and this is ready for another look

File: staging/src/k8s.io/api/apps/v1/types.go
Comment (alculquicondor) [inline staging/src/k8s.io/api/apps/v1/types.go:1]: You can consider adding:

```golang
JobCompletionIndexLabel = JobCompletionIndexAnnotation
```

**Review by alculquicondor — COMMENTED** [review]
/approve
/lgtm

File: staging/src/k8s.io/api/apps/v1/types.go [hunk 9]
```diff
@@ -17,7 +17,7 @@ limitations under the License.
 package v1
 
 import (
-	"k8s.io/api/core/v1"
```
Comment (ahg-g) [inline staging/src/k8s.io/api/apps/v1/types.go:20]: Please revert this file and I can approve this PR.

File: staging/src/k8s.io/api/apps/v1/types.go
Comment (ahg-g) [reply staging/src/k8s.io/api/apps/v1/types.go:1]: lets do that later to not block on an approval on this file.

File: staging/src/k8s.io/api/apps/v1/types.go (hunk 9 above)
Comment (danielvegamyhre) [reply staging/src/k8s.io/api/apps/v1/types.go]: done

**Comment by tengqm** [comment]
Do we have a docs proposal for this new label?

https://kubernetes.io/docs/reference/labels-annotations-taints/

**Comment by danielvegamyhre** [comment]
@tengqm Not yet, that is on my to do list
